RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+#include "RecoEgamma/EgammaTools/interface/EgammaPCAHelper.h"
+
+#include "DataFormats/HGCRecHit/interface/HGCRecHit.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/HcalDetId/interface/HcalDetId.h"
+
+// just to retrieve a "magic" number
+#include "RecoLocalCalo/HGCalRecAlgos/interface/HGCalImagingAlgo.h"
+
+
+#include <algorithm>
+#include <iostream>
+
+EGammaPCAHelper::EGammaPCAHelper(): invThicknessCorrection_({1. / 1.132, 1. / 1.092, 1. / 1.084}),
+                                         pca_(new TPrincipal(3, "D")){
+    hitMapOrigin_ = 0;
+    hitMap_ = new std::map<DetId, const HGCRecHit *>();
+    debug_ = false;
+}
+
+EGammaPCAHelper::~EGammaPCAHelper() {
+    if (hitMapOrigin_ == 2) delete hitMap_;
+}
+void EGammaPCAHelper::setHitMap( std::map<DetId,const HGCRecHit *> * hitMap) {
+    hitMapOrigin_ = 1;
+    hitMap_ = hitMap ;
+    pcaIteration_ = 0;
+}
+
+void EGammaPCAHelper::setRecHitTools(const hgcal::RecHitTools * recHitTools ) {
+    recHitTools_ = recHitTools;
+}
+
+void EGammaPCAHelper::fillHitMap(const HGCRecHitCollection & rechitsEE,
+                                 const HGCRecHitCollection & rechitsFH,
+                                 const HGCRecHitCollection & rechitsBH) {
+    hitMap_->clear();
+    unsigned hitsize = rechitsEE.size();
+    for ( unsigned i=0; i< hitsize ; ++i) {
+        (*hitMap_)[rechitsEE[i].detid()] = & rechitsEE[i];
```

Comment:
for range loop seems more appropriate.
Also, perhaps ```hitMap_->emplace_hint(hitmap_->end(), {rechit.detid(), &rechit});``` may be faster

Reply:
Yeah, this is another heavy spot, I think `unordered_map` might work better, but not sure.

Reply:
IIUC all of the cost paid at insertion is for good: access becomes faster.
My suggestion just tries to benefit from the fact that the vectors are already sorted
(and repeated [i] has some extra cost)